File: Roadmap/04 - CADENAS DE CARACTERES/python/CaveroBrandon.py
"""
Muestra ejemplos de todas las operaciones que puedes realizar con cadenas de caracteres en tu lenguaje.
Algunas de esas operaciones podrían ser (busca todas las que puedas):
- Acceso a caracteres específicos, subcadenas, longitud, concatenación, repetición, recorrido,
conversión a mayúsculas y minúsculas, reemplazo, división, unión, interpolación, verificación
"""


def string_exercise(first_string='First string', second_string='Second string'):
    string_concatenated = first_string + '-' + second_string  # Concatenation of strings
    print('Concatenated string:', string_concatenated)

    string_multiplication = first_string * 3  # Multiplication of a string or repetition
    print('Repetition of a string:', string_multiplication)

    first_char = first_string[0]  # Accessing a specific character
    print('First character of a string:', first_char)

    sliced_string = first_string[0:5]  # Slicing an string
    print('Sliced string:', sliced_string)

    split_string = first_string.split()  # Splitting a string
    print('Split string:', split_string)

    lower_case_string = first_string.lower()  # Transforming any string to lower case
    print('The lower case version of the string is:', lower_case_string)

    upper_case_string = first_string.upper()  # Transforming any string to upper case
    print('The upper case version of the string is:', upper_case_string)

    replace_text = second_string.replace('string', 'text')  # Replacing a string
    print('Replacing the text "string" to "text":', replace_text)

    find_string = first_string.find('string')  # Find a word in a string
    print('The word "string" is in the position:', find_string)

    first_string_length = len(first_string)  # Getting the length of a string
    print('The length of the string is:', first_string_length)

    for i in first_string:  # Navigate a string using a for
        print(i)

    print(f'String interpolation using the {first_string} and the {second_string}')  # String interpolation


"""
DIFICULTAD EXTRA (opcional):
Crea un programa que analice dos palabras diferentes y realice comprobaciones para descubrir si son:
- Palíndromos
- Anagramas
- Isogramas
"""


def is_palindrome(text):
    testing_text = text.lower()
    testing_text = testing_text.replace(' ', '')

    inverted_text = text[::-1]
    inverted_text = inverted_text.lower()
    inverted_text = inverted_text.replace(' ', '')

    if testing_text == inverted_text:
        print(f'The text "{text}" is a palindrome')
    else:
        print(f'The text "{text}" is not a palindrome')


def is_isogram(text):
    test_text = text.replace(' ', '')
    for letter in test_text:
        if test_text.count(letter) != 1:
            print(f'The text "{text}" is not an isogram')
            return
    print(f'The text "{text}" is an isogram')


def is_anagram(text1, text2):
    test_text1 = text1.replace(' ', '')
    test_text1 = test_text1.lower()
    test_text1 = sorted(test_text1)

    test_text2 = text2.replace(' ', '')
    test_text2 = test_text2.lower()
    test_text2 = sorted(test_text2)

    if test_text1 == test_text2:
        print(f'The text {text1} and {text2} are anagram')
    else:
        print(f'The text {text1} and {text2} are not anagram')


string_exercise()

print(' \n **** DIFICULTAD EXTRA ****')
text = input('Input a text to verify if is palindrome and isogram:')
is_palindrome(text)
is_isogram(text)
anagram1 = input('Input a text the first text to verify if is an anagram:')
anagram2 = input('Input a text the first text to verify if is an anagram:')
is_anagram(anagram1, anagram2)

File: Roadmap/04 - CADENAS DE CARACTERES/javascript/CaveroBrandon.js
/*
EJERCICIO:
Muestra ejemplos de todas las operaciones que puedes realizar con cadenas de caracteres
en tu lenguaje. Algunas de esas operaciones podrían ser (busca todas las que puedas):
- Acceso a caracteres específicos, subcadenas, longitud, concatenación, repetición, recorrido,
conversión a mayúsculas y minúsculas, reemplazo, división, unión, interpolación, verificación...
*/
const operationString = 'Hello world';
const supportString = 'Javascript!';

console.log('Original string: ' + operationString);
console.log('Access to second character: ' + operationString[1]);               // Accesing an specific character
console.log('Verifying the existence of a substring: ', operationString.includes('world'))  // Finding a substring
console.log('Getting the substring index: ', operationString.indexOf('world')); // Finding the index of a substring
console.log('Extracting a substring: ', operationString.slice(-5));             // Extracting a substring
console.log('String lenght is: ' + operationString.length);                     // Length of a string
console.log('Concatenated string: ' + `${operationString} - ${supportString}`); // Concatenating strings
console.log('Repeat a string: ', supportString.repeat(2));                      // Repeating a string
console.log('Convert to upper case: ', supportString.toUpperCase());            // Converting to upper cas
console.log('Convert to lower case: ', supportString.toLowerCase());            // Converting to lower case
console.log('Replacing a substring: ', operationString.replace('world', supportString));    // Replacing a substring with other substring
console.log('Spliting a string: ', operationString.split(' '));                 // Spliting a string
console.log(`Interpolating a string in ${supportString}`);                      // Interpolating a string
console.log('Using scape characters:\n', operationString);                      // Scape characters
// Iterating a string
for (i in supportString) {
    console.log(supportString[i]);
}

/*
DIFICULTAD EXTRA (opcional):
Crea un programa que analice dos palabras diferentes y realice comprobaciones
para descubrir si son:
- Palíndromos
- Anagramas
- Isogramas
*/

function preProcess(userText) {
    let lowerCaseString = userText.toLowerCase();
    let originalString = lowerCaseString.split(' ').join('');
    return originalString.split('');
}

function isPalindrome(userText) {
    let lowerCaseString = userText.toLowerCase();
    let originalString = lowerCaseString.split(' ').join('');
    let splitArray = originalString.split('');
    let reversedString = splitArray.reverse().join('');
    
    if (reversedString == originalString) {
        console.log(`The text "${userText}" is palindrome`);
        return true;
    }
    else {
        console.log(`The text "${userText}" is not a palindrome`);
        return false;
    }
}

function isAnagram(firstText, secondText) {
    let firstTextArray = preProcess(firstText);
    let secondTextArray = preProcess(secondText);

    if (firstTextArray.sort().join('') === secondTextArray.sort().join('')) {
        console.log(`The texts "${firstText}" and "${secondText}" are anagrams`);
        return true;
    }
    else {
        console.log(`The texts "${firstText}" and "${secondText}" are not anagrams`);
        return false;
    }

}

function isIsogram(userText) {
    let userTextArray = preProcess(userText);
    
    let evaluationSet = new Set();
    userTextArray.forEach(element => {
        evaluationSet.add(element);
    });
    evaluationArray = Array.from(evaluationSet);
    if (userTextArray.join('') === evaluationArray.join('')) {
        console.log(`The text "${userText}" is an isogram`);
        return true;
    }
    else {
        console.log(`The text "${userText}" is not an isogram`);
        return false;
    }

}

console.log('----- Palindrome -----')
isPalindrome('Hello Javascript');
isPalindrome('Nurses run');
console.log('----- Anagrams -----');
isAnagram('Javascript', 'Cinema');
isAnagram('iceman', 'Cinema');
console.log('----- Isogram -----');
isIsogram('Javascript');
isIsogram('Lumberjacks');
